### ISSUE TITLE
fatal error: snappy.h: No such file or directory

### ISSUE THREAD

**Issue by zhaofenqiang** [PaddlePaddle/Paddle#9291]
My host machine is ubuntu 16.04. And PaddlePaddle for host following [this](http://www.paddlepaddle.org/docs/develop/documentation/zh/howto/capi/compile_paddle_lib_cn.html) can be successfully built.

Now I wanted to build PaddlePaddle for Raspberry Pi following this [guide](https://github.com/PaddlePaddle/Paddle/blob/develop/doc/mobile/cross_compiling_for_raspberry_cn.md).
I used this command to build PaddlePaddle:
```
cmake -DCMAKE_SYSTEM_NAME=RPi  -CMAKE_INSTALL_PREFIX=/home/zfq/Paddle/build/CAPI_ACL_RPi  -DCMAKE_BUILD_TYPE=Debug -DWITH_C_API=ON  -DWITH_SWIG_PY=OFF -DRPI_TOOLCHAIN=/home/zfq/tools/arm-bcm2708/gcc-linaro-arm-linux-gnueabihf-raspbian-x64   -DWITH_GPU=OFF   -DWITH_PYTHON=OFF ..
```

And then encountered the error:
> /home/zfq/Paddle/build/third_party/snappy_stream/src/extern_snappystream/src/osnappystream.cpp:11:20: fatal error: snappy.h: No such file or directory
 #include "snappy.h"

I found the "snappy.h" at `third_party/snappy/src/extern_snappy`, so I modified the `#include` by using absolute path as a workaround. 

But after this I met this error:
> /home/zfq/Paddle/build/third_party/snappy/src/extern_snappy/snappy.h:45:33: fatal error: snappy-stubs-public.h: No such file or directory
 #include "snappy-stubs-public.h"

I also solved it by modified `#include`.

After this I met this:
>/home/zfq/Paddle/paddle/math/MathFunctions.h:39:19: fatal error: cblas.h: No such file or directory
 #include <cblas.h>

Now I am confused, does anyone know why? Maybe the [raspberry_pi.cmake](https://github.com/PaddlePaddle/Paddle/blob/develop/cmake/cross_compiling/raspberry_pi.cmake) has some problem?
Thank you in advance!

**Comment by Xreki** [comment]
[raspberry_pi.cmake](https://github.com/PaddlePaddle/Paddle/blob/develop/cmake/cross_compiling/raspberry_pi.cmake) is only used to set the compiler and flags. All the above errors have no matter about `raspberry_pi.cmake`. I think there is something wrong with the other cmake files.

- Do you want to use Paddle v2 or Fluid? Yet we haven't compiled Fluid on Raspberry Pi. 
- For the first two errors: the dependency of `snappy_stream` is newly added for Fluid. I think there is something wrong with the compiling of `snappy_stream` on Raspberry Pi.
- For the last error: Paddle uses `openblas` default and will automatically download, compile and install `openblas` at `third_party/install/openblas`, do you have this directory?

**Comment by zhaofenqiang** [comment]
@Xreki, Thank you for your reply.
Actually, I am not very clear about the Paddle's version. Just `git clone` and `cmake`...
I am also thinking there's something wrong with `snappy_stream` on  Raspberry Pi.
And I do not have a `openblas` at `third_party/install/`, could you check the code of downloading and compiling `openblas`? 
Thank you very much.

**Comment by Xreki** [comment]
I want to confirm that: **do you want to use Paddle v2 or Fluid on RPi?**

**Comment by zhaofenqiang** [comment]
Paddle v2

**Comment by Xreki** [comment]
Sorry.

- Since you just want to use Paddle v2, I can fix the snappy problem through disable the build of fluid when cross_compiling RPi.

- While for the cblas problem, maybe there is a bug. Would you please paste the output of cmake command here?

**Comment by zhaofenqiang** [comment]
The cmake command is: 
```
cmake -DCMAKE_SYSTEM_NAME=RPi -DRPI_TOOLCHAIN=/home/zfq/tools/arm-bcm2708/arm-rpi-4.9.3-linux-gnueabihf -DRPI_ARM_NEON=ON -DCMAKE_INSTALL_PREFIX=CAPI_RPI -DWITH_GPU=OFF -DWITH_C_API=ON  -DWITH_PYTHON=OFF -DWITH_SWIG_PY=OFF -DCMAKE_BUILD_TYPE=Debug ..
```
And I have made a few changes to the cmake files. 

- modify RPI_TOOLCHAIN from https://github.com/raspberrypi/tools.git to host's default arm-linux-gnueabihf-gcc, and this action probably solved the snappy problem... `fatal error: snappy.h: No such file or directory`nerver happened again.

- set CBLAS_FOUND ON at openblas.cmake's begining, so openblas would be downloaded and installed normally. 

Then the output is: 

> -- Found Paddle host system: ubuntu, version: 16.04
-- Found Paddle host system's CPU: 4 cores
-- Found host C compiler: /usr/bin/cc
-- Found host CXX compiler: /usr/bin/c++
RPI_TOOLCHAIN_PREFIX: /home/zfq/tools/arm-bcm2708/arm-rpi-4.9.3-linux-gnueabihf/bin/arm-linux-gnueabihf-
raw RPI_C_COMPILER: /usr/bin/arm-linux-gnueabihf-gcc
after RPI_C_COMPILER: arm-linux-gnueabihf-gcc
raw RPI_C_COMPILER: /usr/bin/arm-linux-gnueabihf-g++
after RPI_C_COMPILER: arm-linux-gnueabihf-g++
-- CXX compiler: /usr/bin/arm-linux-gnueabihf-g++, version: GNU 5.4.0
-- C compiler: /usr/bin/arm-linux-gnueabihf-gcc, version: GNU 5.4.0
-- Do not have AVX2 intrinsics and disabled MKL-DNN
CMake Warning at /usr/share/cmake-3.11/Modules/FindProtobuf.cmake:455 (message):
  Protobuf compiler version 3.1.0 doesn't match library version 3.2.0
Call Stack (most recent call first):
  cmake/external/protobuf.cmake:17 (FIND_PACKAGE)
  CMakeLists.txt:135 (include)
-- Protobuf protoc executable: /home/zfq/MobileRaw/Paddle/build/third_party/install/protobuf_host/bin/protoc
-- Protobuf library: /home/zfq/MobileRaw/Paddle/build/third_party/install/protobuf/lib/libprotobuf.a
-- Protobuf version: 3.1
-- Found reference-cblas (include: /usr/include, library: /usr/lib/libcblas.so)
CBLAS_FOUND: ON
-- BLAS library: /home/zfq/MobileRaw/Paddle/build/third_party/install/openblas/lib/libopenblas.a
-- warp-ctc library: /home/zfq/MobileRaw/Paddle/build/third_party/install/warpctc/lib/libwarpctc.so
-- Paddle version is 0.11.1a2
-- Skip compiling with MKLDNNMatrix
-- Skip compiling with MKLDNNLayers and MKLDNNActivations
-- Skip compiling with MKLPackedLayers
-- Configuring done
-- Generating done
-- Build files have been written to: /home/zfq/MobileRaw/Paddle/build

For now, I am eager to konw how to disable the build of fluid when cross_compiling for Android or RPi, I only need CAPI, and the build of fluid took up most of the time. Thank you in advance~

**Comment by Xreki** [comment]
- Fluid is disabled for Android, but not disabled for RPi. You may try to add `NOT RPI` to the following cmake files:

https://github.com/PaddlePaddle/Paddle/blob/129859e732fa7ac056c4c453619b2c84c98bc0ac/cmake/external/snappy.cmake#L16-L18

https://github.com/PaddlePaddle/Paddle/blob/129859e732fa7ac056c4c453619b2c84c98bc0ac/cmake/external/snappystream.cmake#L16-L18

https://github.com/PaddlePaddle/Paddle/blob/129859e732fa7ac056c4c453619b2c84c98bc0ac/paddle/CMakeLists.txt#L27-L29

- For the cblas problem, from the cmake output we can see:
>
-- Found reference-cblas (include: /usr/include, library: /usr/lib/libcblas.so)

There should be a `cblas.h` and `libcblas.so` under your system path, but it can be used for rpi. Now please manually set `CBLAS_FOUND` to `OFF` in `openblas.cmake`, we will try to fix this problem.

**Comment by zhaofenqiang** [comment]
All questions solved~
Thank you very much!

**Comment by Xreki** [comment]
I also fix this issue in #9844 . You can have a try. :smile:
